src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.ImmutableTypeParameterArgumentAnalysis.cs

```diff
@@ -0,0 +1,642 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace D2L.CodeStyle.Analyzers.Immutability;
+public sealed partial class ImmutabilityAnalyzer {
+
+	private static class ImmutableTypeParameterArgumentAnalysis {
+
+		public static void Register(
+			CompilationStartAnalysisContext context,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext
+		) {
+			// Type arguments of method return types
+			// Type arguments of method parameters
+			context.RegisterSymbolAction(
+				ctx => {
+					var symbol = (IMethodSymbol)ctx.Symbol;
+
+					// Ignore auto-implemened property methods
+					switch( symbol.MethodKind ) {
+						case MethodKind.PropertyGet:
+						case MethodKind.PropertySet:
+							return;
+					}
+
+					SyntaxNode getBaseSyntax() => symbol.DeclaringSyntaxReferences[ 0 ].GetSyntax( ctx.CancellationToken );
+
+					if( symbol.MethodKind != MethodKind.Constructor ) {
+						AnalyzeTypeRecursive(
+							ctx.ReportDiagnostic,
+							annotationsContext,
+							immutabilityContext,
+							symbol.ReturnType,
+							() => {
+								SyntaxNode syntax = getBaseSyntax();
+								return getBaseSyntax() switch {
+									MethodDeclarationSyntax methodDeclaration => methodDeclaration.ReturnType,
+									_ => syntax,
+								};
+							}
+						);
+					}
+
+					for( int i = 0; i < symbol.Parameters.Length; ++i ) {

```

Comment:
nit: `i++` is idiomatic in C#. `++i` is a C++ thing where it is common to overload `++` for iterators and the `i++` will involve copying the iterator, but that's not a thing in C#.

Reply:
Yeah, just habit. :+1: